shotover-proxy/src/transforms/cassandra/sink_cluster/token_map.rs

```diff
@@ -0,0 +1,109 @@
+pub struct TokenMap {
+    token_ring: BTreeMap<Murmur3Token, Uuid>,
+}
+
+impl TokenMap {
+    pub fn new(nodes: &[CassandraNode]) -> Self {
+        TokenMap {
+            token_ring: nodes
+                .iter()
+                .flat_map(|node| node.tokens.iter().map(move |token| (*token, node.host_id)))
+                .collect(),
+        }
+    }
+
+    /// Returns nodes starting at given token and going in the direction of replicas.
+    pub fn iter_replica_nodes(
+        &self,
+        token: Murmur3Token,
+        replica_count: usize,
+    ) -> impl Iterator<Item = Uuid> + '_ {
+        self.token_ring
+            .range(token..)
+            .chain(self.token_ring.iter())
+            .take(replica_count)
+            .map(|(_, node)| *node)
+    }
+}
+
+#[cfg(test)]
+mod test_token_map {
+    use super::*;
+    use hex_literal::hex;
+    use itertools::Itertools;
+    use uuid::Uuid;
+
+    static NODE_1: Uuid = Uuid::from_bytes(hex!("2DD022D62937475489D602D2933A8F71"));
+    static NODE_2: Uuid = Uuid::from_bytes(hex!("2DD022D62937475489D602D2933A8F72"));
+    static NODE_3: Uuid = Uuid::from_bytes(hex!("2DD022D62937475489D602D2933A8F73"));
+
+    fn prepare_nodes() -> Vec<CassandraNode> {
+        vec![
+            CassandraNode::new(
+                "127.0.0.1:9042".parse().unwrap(),
+                "dc1".into(),
+                vec![
+                    Murmur3Token::new(-2),
+                    Murmur3Token::new(-1),
+                    Murmur3Token::new(0),
+                ],
+                NODE_1,
+            ),
+            CassandraNode::new(
+                "127.0.0.1:9043".parse().unwrap(),
+                "dc1".into(),
+                vec![Murmur3Token::new(20)],
+                NODE_2,
+            ),
+            CassandraNode::new(
+                "127.0.0.1:9044".parse().unwrap(),
+                "dc1".into(),
+                vec![
+                    Murmur3Token::new(2),
+                    Murmur3Token::new(1),
+                    Murmur3Token::new(10),
+                ],
+                NODE_3,
+            ),
+        ]
+    }
+
+    #[test]
+    fn should_return_replicas_in_order() {
+        verify_tokens(
+            &[NODE_1, NODE_3, NODE_3, NODE_3, NODE_2],
+            Murmur3Token::new(0),
+        );
+    }
+
+    #[test]
+    fn should_return_replicas_in_order_for_non_primary_token() {
+        verify_tokens(&[NODE_3, NODE_2], Murmur3Token::new(3));
+    }
+
+    #[test]
+    fn should_return_replicas_in_a_ring() {
+        verify_tokens(
+            &[NODE_2, NODE_1, NODE_1, NODE_1, NODE_3],
+            Murmur3Token::new(20),
+        );
+    }
+
+    fn verify_tokens(node_host_ids: &[Uuid], token: Murmur3Token) {
+        let token_map = TokenMap::new(prepare_nodes().as_slice());
+        let nodes = token_map
+            .iter_replica_nodes(token, node_host_ids.len())
+            .collect_vec();
+
+        assert_eq!(nodes.len(), node_host_ids.len());
```

Comment:
I believe we can simplify these assertions to just `assert_eq!(nodes, node_host_ids);`.
It should give better error messages too!